types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,645 @@
+				"given output");
+		return;
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector = NULL, *tmp_conn;
+	wl_list_for_each(tmp_conn, &device->connectors, link) {
+		if (tmp_conn->output == output) {
+			connector = tmp_conn;
+			break;
+		}
+	}
+
+	if (!connector) {
+		wlr_log(WLR_DEBUG, "Output has not been offered");
+		return;
+	}
+
+	if (connector->active_lease) {
+		wlr_drm_lease_v1_revoke(connector->active_lease);
+	}
+}
+
+static void drm_lease_device_v1_destroy(
+		struct wlr_drm_lease_device_v1 *device) {
+	struct wlr_drm_backend *backend =
+			get_drm_backend_from_backend(device->backend);
+	wlr_log(WLR_DEBUG, "Destroying wlr_drm_lease_device_v1 for %s",
+			backend->name);
+
+	if (!device) {
+		return;
+	}
+
+	struct wl_resource *resource;
+	struct wl_resource *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &device->resources) {
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_resource_set_user_data(resource, NULL);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->requests) {
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_resource_set_user_data(resource, NULL);
+	}
+
+	wl_resource_for_each_safe(resource, tmp_resource, &device->leases) {
+		struct wlr_drm_lease_v1 *lease = drm_lease_v1_from_resource(resource);
+		wlr_drm_lease_v1_revoke(lease);
+
+		wl_list_remove(wl_resource_get_link(resource));
+		wl_resource_set_user_data(resource, NULL);
+	}
+
+	struct wlr_drm_lease_connector_v1 *connector, *tmp;
+	wl_list_for_each_safe(connector, tmp, &device->connectors, link) {
+		wlr_drm_lease_v1_manager_withdraw_output(device->manager,
+				connector->output);
+	}
+
+	wl_list_remove(&device->link);
+	wlr_global_destroy_safe(device->global, device->manager->display);
+
+	free(device);
+}
+
+static void handle_backend_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_drm_lease_device_v1 *device =
+		wl_container_of(listener, device, backend_destroy);
+
+	drm_lease_device_v1_destroy(device);
+}
+
+static void drm_lease_device_v1_create(struct wlr_drm_lease_v1_manager *manager,
+		struct wlr_backend *backend) {
+	assert(backend);
+
+	struct wlr_drm_backend *drm_backend =
+			get_drm_backend_from_backend(backend);
+	wlr_log(WLR_DEBUG, "Creating wlr_drm_lease_device_v1 for %s",
+			drm_backend->name);
+
+	struct wlr_drm_lease_device_v1 *lease_device =
+		calloc(1, sizeof(struct wlr_drm_lease_device_v1));
+
+	if (!lease_device) {
+		wlr_log(WLR_ERROR, "Failed to allocate wlr_drm_lease_device_v1");
+		return;
+	}
+
+	lease_device->manager = manager;
+	lease_device->backend = backend;
+	wl_list_init(&lease_device->resources);
+	wl_list_init(&lease_device->connectors);
+	wl_list_init(&lease_device->requests);
+	wl_list_init(&lease_device->leases);
+	wl_list_init(&lease_device->link);
+
+	lease_device->backend_destroy.notify = handle_backend_destroy;
+
+	wl_signal_add(&backend->events.destroy, &lease_device->backend_destroy);
```

Comment:
Need to add the signal after `wl_global_create`. Otherwise, on `wl_global_create` failure, we'll have a dangling listener that points to free'd memory.